File: src/main/java/com/andredonadon/sorting/Merge.java
package com.andredonadon.sorting;

import java.util.Arrays;

public class Merge {

    public static void main(String[] args) {
        int[] array = {12, 15, 7, 9, 2, 5, 1, 10, 4, 8};

        System.out.println("Original array:" + Arrays.toString(array));
        mergeSort(array, 0, array.length - 1);
        System.out.println("Ordered array:" + Arrays.toString(array));
    }

    private static void merge(int[] A, int p, int q, int r) {
        int n1 = q - p + 1;
        int n2 = r - q;

        int[] L = new int[n1 + 2];
        int[] R = new int[n2 + 2];

        for (int i = 0; i < n1; i++) {
            L[i] = A[p + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = A[q + 1 + j];
        }

        L[n1] = Integer.MAX_VALUE;
        R[n2] = Integer.MAX_VALUE;

        int i = 0, j = 0;

        for (int k = p; k <= r; k++) {
            if (L[i] <= R[j]) {
                A[k] = L[i];
                i++;
            } else {
                A[k] = R[j];
                j++;
            }
        }
    }

    private static void mergeSort(int[] A, int p, int r) {
        if (p < r) {
            int q = (p + r) / 2;
            mergeSort(A, p, q);
            mergeSort(A, q + 1, r);
            merge(A, p, q, r);
        }
    }

}

File: kalgorithms/src/main/kotlin/com/andredonadon/kalgorithms/sorting/Merge.kt
package com.andredonadon.kalgorithms.sorting

fun mergeSort(arr: IntArray, p: Int, r: Int) {
    if (p < r) {
        val q = (p + r) / 2
        mergeSort(arr, p, q)
        mergeSort(arr, q + 1, r)
        merge(arr, p, q, r)
    }
}

fun merge(arr: IntArray, p: Int, q: Int, r: Int) {
    val n1 = q - p + 1
    val n2 = r - q

    val L = IntArray(n1 + 1)
    val R = IntArray(n2 + 1)

    for (i in 0 until n1) {
        L[i] = arr[p + i]
    }
    for (j in 0 until n2) {
        R[j] = arr[q + 1 + j]
    }

    L[n1] = Int.MAX_VALUE
    R[n2] = Int.MAX_VALUE

    var i = 0
    var j = 0

    for (k in p..r) {
        if (L[i] <= R[j]) {
            arr[k] = L[i]
            i++
        } else {
            arr[k] = R[j]
            j++
        }
    }
}

fun main() {
    val array = intArrayOf(12, 15, 7, 9, 2, 5, 1, 10, 4, 8)
    println("Array original: ${array.joinToString(", ")}")
    mergeSort(array, 0, array.size - 1)
    println("Array ordenado: ${array.joinToString(", ")}")
}